File: round_4/D_Military_Problem.cpp
#include <bits/stdc++.h>

using namespace std;


int spread(vector<vector<int>>& graph, vector<int>& allAnswer, vector<int>& child_count, int node, int parent){
    int children = 1;
    allAnswer.push_back(node);
    for (auto child : graph[node]){
        if (child != parent){
            children += spread(graph, allAnswer, child_count, child, node);
        }
    }
    child_count[node] = children;
    return children;
}




int main (){
    int n, q; 
    cin >> n >> q;
    vector<int> allAnswer;
    vector<vector<int>> graph(n, vector<int>());
    vector<int> child_count(n, 0);
    vector<int> answer(q, -1);
    for (int i = 0; i < n - 1; ++i){
        int child;
        cin >> child;
        graph[child - 1].push_back(i + 1);
        graph[i + 1].push_back(child - 1);
    }

    for (vector<int> &v : graph){
        sort(v.begin(), v.end());
    }

    spread(graph, allAnswer, child_count, 0, -1);
    unordered_map<int, int> map;
    for (int i = 0; i < allAnswer.size(); i ++){
        map[allAnswer[i]] = i;
    }
    
    for (int i = 0; i < q; i ++ ){
        int u, k;
        cin >> u >> k;
        u --;
        k --;
        int start_index = map[u];
        // cout << k;
        int end_k = start_index + k;
        // cout << "start_index: " << start_index << " end_k: " << end_k << endl;
        if (k < child_count[u]){
            answer[i] = allAnswer[end_k] + 1;
        } 
        cout << answer[i] << endl;
    }


    




}

File: round_4/D_Military_Problem.py

import sys, threading

input = lambda: sys.stdin.readline().strip()

def main():
    n, q = [int(i) for i in input().split()]
    graph = [[] for i in range(n)]
    edges = [int(i) for i in input().split()]
    child_count = [0 for i in range(n)]
    for i in range(len(edges)):
        graph[i + 1].append(edges[i]-1)
        graph[edges[i]-1].append(i + 1)
    for i in range(len(graph)):
        graph[i].sort()
    
    answer = []

    def dfs(graph, node, parent):
        nonlocal answer
        answer.append(node)
        children = 1
        for child in graph[node]:
            if child != parent:
                children += dfs(graph, child, node)
        child_count[node] = children
        return children
    dfs(graph, 0, -1)


    m = {}
    for i in range(len(answer)):
        m[answer[i]] = i
    result = [-1 for i in range(q)]
    # print(answer)
    for i in range(q):
        u, k = [int(i) - 1 for i in input().split()]
        start_index = m[u]
        end_k = start_index + k 
        
        if k < child_count[u]:
            result[i] = answer[end_k] + 1
    
    print(*result, sep="\n")
    
if __name__ == '__main__':
    
    sys.setrecursionlimit(1 << 30)
    threading.stack_size(1 << 27)

    main_thread = threading.Thread(target=main)
    main_thread.start()
    main_thread.join()
